client/device/cloudftd/upgrade.go

```diff
@@ -46,11 +49,52 @@ func (f *ftdUpgradeService) Upgrade(uid string, softwareVersion string) (*FtdDev
 	if err != nil {
 		return nil, err
 	}
-	err = f.validateFtdVersion(ftdDevice, softwareVersion)
+	err = f.validateFtdVersion(ftdDevice, softwareVersionStr)
 	if err != nil {
 		return nil, err
 	}
 
+	tflog.Debug(f.Ctx, "Triggering the FTD device upgrade.")
+	return f.doUpgrade(ftdDevice, softwareVersionStr)
+}
+
+func (f *ftdUpgradeService) doUpgrade(ftdDevice *FtdDevice, softwareVersionStr string) (*FtdDevice, error) {
```

Comment:
See my comments below; we're already fetching all upgrade packages before this; make the changes suggested there and pass in the upgradePackage as the second argument to this method